### ISSUE TITLE
ENOSPC-Errors on startup of Zettlr

### ISSUE THREAD

**Issue by nathanlesage** [Zettlr/Zettlr#14]
Hey,

I've now received error reports from multiple persons that relate to so-called ENOSPC-errors; mainly on Linux systems, but there were instances of these errors occurring on Windows (thanks to @damien122 for reporting this in issue #13).

While I was always unsure as to why this happens, because I never experienced these errors on any platform, and they seemed to be completely random, I now know why they occur.

The errors are related to the hard-coded maximum amount of file handles that can be opened simultaneously by an app on each operating system. As has been written in [this thread on StackOverflow](https://stackoverflow.com/questions/16748737/grunt-watch-error-waiting-fatal-error-watch-enospc/17437601#17437601), one could manually increase the amount of files.

But as this is unfeasible for all systems, and I don't want to mess with environment variables, there are two options:

1. Simply notifying the user that the current directory simply contains too many files for the `chokidar`-plugin to use and that the directory will therefore not be monitored and external changes cannot be reflected by the app.
2. Which I deem the better one: Throwing out `chokidar` in favor of a complex implementation of "pulses" that are sent through the directory trees every like 30 seconds to re-scan the contents and reflect potential changes and simply `fs.watch()` the currently opened file directly.

Any thoughts on that?

**Comment by damien122** [comment]
Hi,
IMHO:
how about to show small window on first startup to let the user define the "Zettlr home" directory. This should be an empty directory only for Zettlr and could be also stored in an OneDrive or Google Drive synced folder, to keep the files backuped.
Of cource such a "Zettlr home" directory could be not empty and could also lead to the same bug, but this could be catched with some user hints.
Also this path could be changeable in the Zettlr settings dialog for later changes.
This would be an easy solution and lead to more usability.

**Comment by nathanlesage** [comment]
I see where you are going to, but maybe an easier idea for the people (nowadays, everybody expects apps to ask as little questions as possible, and somehow I think it's okay for this specific use case): 

Instead of always defaulting to the complete Documents folder, which may be littered with files, why not use the approach some other apps are also using: Simply fallback to a new `zettlr-files`-Folder in the User's Documents directory? This would ensure that it's empty on first startup, making it less error-prone. And on subsequent starts, because if a folder only consists of folders and markdown files, Zettlr will be able to handle it (I believe nobody on this planet has more than 20.000 markdown files in a single root directory).

I would think that this is a good compromise between comfortability, ease of implementation and security.

**Comment by damien122** [comment]
yes it sounds good to create a new folder in documents folder only for zettlr. But then it should be possible to add more root folder or change this folder.

**Comment by nathanlesage** [comment]
@damien122 You can always change the root folder, by pressing Ctrl+O, the first toolbar button or selecting the appropriate option from the menu.

Concerning multiple root folders, see issue #3

Another friend from me has suggested to simply ignore dotted folders, as they are dotted for some reason, namely that they don't contain any data of importance for the purposes of Zettlr.

I think I will go and work on this the next days, so that we can at least sleep safe again :D

**Comment by nathanlesage** [comment]
@damien122 Would you like to test the current develop branch of Zettlr again against your Documents directory to check whether or not this issue still exists?

Also, I'm open for suggestions as to which kinds of directories (except the already excluded Dot-directories and *.app-directories on macOS) should be added to the exclusion list by default!

**Comment by damien122** [comment]
@nathanlesage yes shure I can test ist, but just at Monday because this bug was on my work pc.

**Comment by nathanlesage** [comment]
Alright, thanks a lot! :)

(Oh, and please beware: I just started working on the Autosave/Revert-feature mentioned in #7, it may be that there are some unrelated errors, but they should be pretty obvious — I'm just telling this because I don't know whether or not I can manage to completely implement the feature until monday, so please don't be surprised! And yes, for the next features I'll create additional branches!)

**Comment by damien122** [comment]
Hi @nathanlesage,

I have testet the dev. branch and it works fine. No crashes any more. But what I don't like is, that my full Documents folder is parsed and showen in the navigation pannel. With a lot of empty folders. Or with stuff which is not a "note" like my ssh key or such stuff.
I think it would be better to create an empty folder just for zettlr at first startup and use this as home folder.

Regards,
damien122

**Comment by nathanlesage** [comment]
Hey @damien122!

Thanks for testing this out! So now I'll be closing this issue, because the other thing you've mentioned (to not completely parse the documents folder) can easily be resolved through the implementation of issue #3. Because, if I enable Zettlr to hold multiple root folders, I can solve several issues at once:

1. Open several root folders (obviously)
2. Open no folder on first startup at all (e.g. simply present the user with a little plus sign or something like that to indicate that he/she can begin adding stuff)
3. Also open single files. This would then enable the app to register itself as handler for *.md-files, so that you can double click on those and open single files without parsing a whole folder.

This will take some time, but I think it'll improve the app experience a lot!
